// solid-app-router/dist/integration.js
import { createSignal, onCleanup } from "solid-js";
function bindEvent(target, type, handler) {
    target.addEventListener(type, handler);
    return () => target.removeEventListener(type, handler);
}
function intercept([value, setValue], get, set) {
    return [get ? () => get(value()) : value, set ? (v) => setValue(set(v)) : setValue];
}
export function createIntegration(get, set, init, utils) {
    let ignore = false;
    const signal = intercept(createSignal({ value: get() }, { equals: (a, b) => a.value === b.value }), undefined, next => {
        !ignore && set(next);
        return next;
    });
    init &&
        onCleanup(init((value = get()) => {
            ignore = true;
            signal[1]({ value });
            ignore = false;
        }));
    return {
        signal,
        utils
    };
}
export function normalizeIntegration(integration) {
    if (!integration) {
        return {
            signal: createSignal({ value: "" })
        };
    }
    else if (Array.isArray(integration)) {
        return {
            signal: integration
        };
    }
    return integration;
}
export function staticIntegration(obj) {
    return {
        signal: [() => obj, next => Object.assign(obj, next)]
    };
}
export function pathIntegration() {
    return createIntegration(() => window.location.pathname + window.location.search, ({ value, replace }) => {
        if (replace) {
            window.history.replaceState(null, "", value);
        }
        else {
            window.history.pushState(null, "", value);
        }
        window.scrollTo(0, 0);
    }, notify => bindEvent(window, "popstate", () => notify()));
}
export function hashIntegration() {
    return createIntegration(() => window.location.hash.slice(1), ({ value }) => {
        window.location.hash = value;
        window.scrollTo(0, 0);
    }, notify => bindEvent(window, "hashchange", () => notify()), {
        renderPath: path => `#${path}`
    });
}
